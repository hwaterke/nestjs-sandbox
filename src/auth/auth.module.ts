import {Module} from '@nestjs/common'
import {AuthService} from './auth.service'
import {AuthResolver} from './auth.resolver'
import {User} from '../users/entities/user.entity'
import {TypeOrmModule} from '@nestjs/typeorm'
import {JwtModule} from '@nestjs/jwt'
import {PassportModule} from '@nestjs/passport'
import {JwtStrategy} from './jwt.strategy'

@Module({
  imports: [
    PassportModule,
    JwtModule.register({
      secret: process.env.JWT_SECRET ?? 's3cr3t',
    }),
    TypeOrmModule.forFeature([User]),
  ],
  providers: [AuthResolver, AuthService, JwtStrategy],
  exports: [JwtStrategy],
})
export class AuthModule {}
