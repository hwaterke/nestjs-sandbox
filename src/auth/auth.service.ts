import {Injectable} from '@nestjs/common'
import {RegisterInput} from './dto/register.input'
import {UserObjectType} from '../users/types/user.type'
import {Repository} from 'typeorm'
import {User} from '../users/entities/user.entity'
import {InjectRepository} from '@nestjs/typeorm'
import * as bcrypt from 'bcrypt'
import {LoginInput} from './dto/login.input'

@Injectable()
export class AuthService {
  constructor(
    @InjectRepository(User) private userRepository: Repository<User>
  ) {}

  async register(registerInput: RegisterInput): Promise<UserObjectType> {
    const emailAvailable = await this.isEmailAvailable(registerInput.email)
    if (!emailAvailable) {
      throw new Error('EMAIL_ALREADY_USED')
    }

    const hashedPassword = await this.hashPassword(registerInput.password)

    const user = this.userRepository.create({
      email: registerInput.email,
      password: hashedPassword,
    })

    return this.userRepository.save(user)
  }

  async login(loginInput: LoginInput) {
    const user = await this.userRepository.findOne({email: loginInput.email})

    if (user) {
      const passwordMatch = await this.checkPassword(
        loginInput.password,
        user.password
      )
      if (passwordMatch) {
        return user
      }
    }
    throw new Error('INCORRECT_EMAIL_OR_PASSWORD')
  }

  hashPassword(password: string) {
    return bcrypt.hash(password, 12)
  }

  checkPassword(password: string, hashedPassword: string) {
    return bcrypt.compare(password, hashedPassword)
  }

  async isEmailAvailable(email: string) {
    const totalEmails = await this.userRepository.count({email})
    return totalEmails === 0
  }
}
