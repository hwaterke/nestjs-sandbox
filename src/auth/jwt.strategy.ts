import {Injectable, UnauthorizedException} from '@nestjs/common'
import {PassportStrategy} from '@nestjs/passport'
import {InjectRepository} from '@nestjs/typeorm'
import {ExtractJwt, Strategy} from 'passport-jwt'
import {Repository} from 'typeorm'
import {User} from '../users/entities/user.entity'
import {JwtPayload} from './jwt.payload.interface'

@Injectable()
export class JwtStrategy extends PassportStrategy(Strategy) {
  constructor(
    @InjectRepository(User) private userRepository: Repository<User>
  ) {
    super({
      jwtFromRequest: ExtractJwt.fromAuthHeaderAsBearerToken(),
      secretOrKey: process.env.JWT_SECRET ?? 's3cr3t',
    })
  }

  async validate(payload: JwtPayload): Promise<User> {
    const user = await this.userRepository.findOne(payload.uuid)

    if (!user) {
      throw new UnauthorizedException()
    }

    return user
  }
}
