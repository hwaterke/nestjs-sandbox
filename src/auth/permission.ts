export enum Permission {
  CREATE_USERS = 'CREATE_USERS',
  READ_USERS = 'READ_USERS',
}
