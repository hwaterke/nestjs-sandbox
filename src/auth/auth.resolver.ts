import {Args, Mutation, Resolver} from '@nestjs/graphql'
import {AuthService} from './auth.service'
import {RegisterInput} from './dto/register.input'
import {UserObjectType} from '../users/types/user.type'

@Resolver()
export class AuthResolver {
  constructor(private readonly authService: AuthService) {}

  @Mutation(() => UserObjectType)
  register(@Args('registerInput') registerInput: RegisterInput) {
    return this.authService.register(registerInput)
  }
}
