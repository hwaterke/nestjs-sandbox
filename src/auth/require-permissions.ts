import {SetMetadata} from '@nestjs/common'
import {Permission} from './permission'

export const PERMISSIONS_METADATA_KEY = 'permissions'

export const RequirePermissions = (...permissions: Permission[]) =>
  SetMetadata(PERMISSIONS_METADATA_KEY, permissions)
