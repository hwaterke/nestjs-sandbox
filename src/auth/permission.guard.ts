import {CanActivate, ExecutionContext, Injectable} from '@nestjs/common'
import {Reflector} from '@nestjs/core'
import {GqlExecutionContext} from '@nestjs/graphql'
import {Permission} from './permission'
import {PERMISSIONS_METADATA_KEY} from './require-permissions'

@Injectable()
export class PermissionGuard implements CanActivate {
  constructor(private reflector: Reflector) {}

  canActivate(context: ExecutionContext): boolean {
    const ctx = GqlExecutionContext.create(context)

    // Read the metadata
    const requiredPermissions = this.reflector.getAllAndOverride<Permission[]>(
      PERMISSIONS_METADATA_KEY,
      [ctx.getHandler(), ctx.getClass()]
    )

    if (!requiredPermissions) {
      return true
    }

    // eslint-disable-next-line @typescript-eslint/no-unsafe-member-access
    const {user} = ctx.getContext().req

    // TODO - Check if the user has the right permissions

    return true
  }
}
