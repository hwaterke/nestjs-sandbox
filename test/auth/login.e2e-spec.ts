import {Test, TestingModule} from '@nestjs/testing'
import {gql} from 'apollo-server-express'
import {Connection} from 'typeorm'
import {AppModule} from '../../src/app.module'
import {AuthService} from '../../src/auth/auth.service'
import {TestingUtils} from '../utils/testing.utils'

const LOGIN = gql`
  mutation login($email: String!, $password: String!) {
    login(email: $email, password: $password) {
      user {
        email
      }
      token
    }
  }
`

describe('AuthResolver.login (e2e)', () => {
  let connection: Connection
  let authService: AuthService
  let testUtils: TestingUtils

  beforeAll(async () => {
    const moduleFixture: TestingModule = await Test.createTestingModule({
      imports: [AppModule],
    }).compile()

    testUtils = await TestingUtils.initNestApplication(moduleFixture)

    connection = moduleFixture.get(Connection)
    authService = moduleFixture.get(AuthService)
  })

  afterEach(() => {
    testUtils.token = null
    return connection.synchronize(true)
  })

  afterAll(() => {
    return testUtils.close()
  })

  it('should fail with an non existing email', async () => {
    const result = await testUtils.apolloClient.mutate({
      mutation: LOGIN,
      variables: {
        email: 'test@example.com',
        password: '12345678',
      },
    })

    expect(result).toMatchSnapshot()
  })

  it('should return the current user and a token', async () => {
    await authService.register({
      email: 'test@example.com',
      password: '12345678',
    })

    const result = await testUtils.apolloClient.mutate({
      mutation: LOGIN,
      variables: {
        email: 'test@example.com',
        password: '12345678',
      },
    })

    expect(result.errors).toBeUndefined()
    // eslint-disable-next-line @typescript-eslint/no-unsafe-member-access
    expect(result.data.login.user).toMatchSnapshot()
    // eslint-disable-next-line @typescript-eslint/no-unsafe-member-access
    expect(result.data.login.token).toBeDefined()
  })
})
