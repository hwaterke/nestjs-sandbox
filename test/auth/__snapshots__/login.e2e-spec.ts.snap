// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AuthResolver.login (e2e) should fail with an non existing email 1`] = `
Object {
  "data": null,
  "errors": Array [
    [GraphQLError: INCORRECT_EMAIL_OR_PASSWORD],
  ],
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`AuthResolver.login (e2e) should return the current user and a token 1`] = `
Object {
  "email": "test@example.com",
}
`;
